pallets/asset/src/lib.rs

```diff
@@ -352,6 +352,40 @@ decl_module! {
         const AssetNameMaxLength: u32 = T::AssetNameMaxLength::get();
         const FundingRoundNameMaxLength: u32 = T::FundingRoundNameMaxLength::get();
 
+        fn on_runtime_upgrade() -> frame_support::weights::Weight {
+            use frame_support::weights::constants::WEIGHT_PER_MICROS;
+            // Keep track of upgrade cost.
+            let mut weight = 0u64;
+            storage_migrate_on!(StorageVersion::get(), 1, {
+                let mut total_len = 0u64;
+                // Get list of assets with invalid asset_types.
+                let fix_list = Tokens::iter()
+                    .filter(|(_, token)| {
+                        total_len += 1;
+                        // Check if the asset_type is invalid.
+                        Self::ensure_asset_type_valid(token.asset_type).is_err()
+                    }).map(|(ticker, _)| ticker).collect::<Vec<_>>();
```

Comment:
Not sure why this allocation is necessary. You should be able to mutate the asset type on the fly instead. with the mutation in an `.inspect(...)` and then a `.count()` at the end to compute `fix_len`.

Reply:
The docs for `iter` says that not to mutate the map during it iteration: https://paritytech.github.io/substrate/latest/frame_support/storage/types/struct.StorageMap.html#method.iter

Reply:
The migration code is fine as is.  I have tested it with 100 tokens.

Reply:
Oh yeah, you're right, although I wonder if that's actually true. For example, consider this implementation of `translate`: https://paritytech.github.io/substrate/latest/src/frame_support/storage/generator/map.rs.html#123-209 -- clearly it's mutating the map during iteration, but it is OK, because there's no change to the keys in the map. So ostensibly, what they really mean by the comment there is "If you change what the keys hash to, you get undefined results."

Reply:
I wasn't sure if the docs only meant adding/removing keys or also value changes.  I would think that normally mutating the value shouldn't, but we can't be sure with a key/value store like RocksDb.  Or if the key/value is replaced sometime in the future.  The way the storage engine iterates might change when doing a store, even if none of the keys change.

I would have preferred to do an inplace mutate, the closest is `translate`, but that will do a store for all values even if the value doesn't change.

So I played it safe with this code.

Reply:
Yep, makes sense 👍 -- might be worth getting docs clarification from upstream in the future.